cfgmgr/coppmgr.cpp

```diff
@@ -0,0 +1,745 @@
+            else if (table_name == CFG_COPP_GROUP_TABLE_NAME)
+            {
+                m_coppGroupInitCfg[table_key] = fvs;
+            }
+        }
+    }
+}
+
+/* Check if the trap group has traps that can be installed only when
+ * feature is enabled
+ */
+bool CoppMgr::checkTrapGroupPending(string trap_group_name)
+{
+    bool traps_present = false;
+    for (auto it: m_coppTrapIdTrapGroupMap)
+    {
+        if (it.second == trap_group_name)
+        {
+            traps_present = true;
+            /* At least one trap should be enabled to install the trap group
+             */
+            if (!isTrapIdDisabled(it.first))
+            {
+                return false;
+            }
+        }
+    }
+    return traps_present;
+}
+
+/* Feature name and CoPP Trap table name must match */
+void CoppMgr::setFeatureTrapIdsStatus(string feature, bool enable)
+{
+    bool disabled_trap = (m_coppDisabledTraps.find(feature)  != m_coppDisabledTraps.end());
+
+    if ((enable && !disabled_trap) || (!enable && disabled_trap))
+    {
+        return;
+    }
+
+    if (m_coppTrapConfMap.find(feature) == m_coppTrapConfMap.end())
+    {
+        if (!enable)
+        {
+            m_coppDisabledTraps.insert(feature);
+        }
+        return;
+    }
+    string trap_group = m_coppTrapConfMap[feature].trap_group;
+    bool prev_group_state = checkTrapGroupPending(trap_group);
+
+    if (!enable)
+    {
+        m_coppDisabledTraps.insert(feature);
+    }
+    else
+    {
+        m_coppDisabledTraps.erase(feature);
+    }
+
+    /* Trap group moved to pending state when feature is disabled. Remove trap group
+    */
+    if (checkTrapGroupPending(trap_group) && !prev_group_state)
+    {
+        m_appCoppTable.del(trap_group);
+        delCoppGroupStateOk(trap_group);
+        return;
+    }
+    vector<FieldValueTuple> fvs;
+    string trap_ids;
+
+    /* Trap group moved from pending state to enabled state
+     * Hence install include all fields to create the group
+     */
+    if (prev_group_state && !checkTrapGroupPending(trap_group))
+    {
+        for (auto i : m_coppGroupFvs[trap_group])
+        {
+            FieldValueTuple fv(i.first, i.second);
+            fvs.push_back(fv);
+        }
+    }
+    getTrapGroupTrapIds(trap_group, trap_ids);
+    if (!trap_ids.empty())
+    {
+        FieldValueTuple fv(COPP_TRAP_ID_LIST_FIELD, trap_ids);
+        fvs.push_back(fv);
+    }
+    if (!fvs.empty())
+    {
+        m_appCoppTable.set(trap_group, fvs);
+        setCoppGroupStateOk(trap_group);
+    }
+}
+
+bool CoppMgr::isTrapIdDisabled(string trap_id)
+{
+    for (auto &m: m_coppDisabledTraps)
+    {
+        if (m_coppTrapConfMap.find(m) == m_coppTrapConfMap.end())
+        {
+            continue;
+        }
+        vector<string> trap_id_list;
+
+        trap_id_list = tokenize(m_coppTrapConfMap[m].trap_ids, list_item_delimiter);
+        if(std::find(trap_id_list.begin(), trap_id_list.end(), trap_id) != trap_id_list.end())
```

Comment:
m_coppDisabledTraps is not trap ids but COPP_TRAP key. This will contain list of trap ids. Thus to find if a trap id is disabled, we need to go through the list of disabled traps and iterate through their trap ids